### ISSUE TITLE
Loading model error 'No such file or directory: 'model.weights.h5'

### ISSUE THREAD

**Issue by WeichenXu123** [tensorflow/tensorflow#70273]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.18.0-dev20240623

### Custom code

No

### OS platform and distribution

Ubuntu 22.04

### Python version

3.9

### Current behavior?

If I launches multiple concurrency processes to load tensorflow model from one saved model file, error occurs like:

```
  File "/tmp/tfm2/t1.py", line 13, in <module>
    tensorflow.keras.models.load_model(model_file_path, custom_objects=custom_objects)
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_api.py", line 182, in load_model
    return saving_lib.load_model(
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_lib.py", line 229, in load_model
    return _load_model_from_fileobj(
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_lib.py", line 353, in _load_model_from_fileobj
    weights_file_path.unlink()
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/pathlib.py", line 1354, in unlink
    self._accessor.unlink(self)

```

### Standalone code to reproduce the issue

```shell
(1) Create the following python script `tf-save.py` to generate model file:


import os.path

import pandas as pd
import numpy as np
from sklearn import datasets
from tensorflow.keras.layers import Concatenate, Dense, Input, Lambda
from tensorflow.keras.saving import register_keras_serializable
from tensorflow.keras.models import Model, Sequential
from tensorflow.keras.optimizers import SGD
import cloudpickle
import sys

save_path = sys.argv[1]

iris = datasets.load_iris()
data = pd.DataFrame(
    data=np.c_[iris["data"], iris["target"]], columns=iris["feature_names"] + ["target"]
)
y = data["target"]
x = data.drop("target", axis=1)

input_a = Input(shape=(2, 3), name="a")
input_b = Input(shape=(2, 5), name="b")


@register_keras_serializable(name="f2")
def f2(z):
    from tensorflow.keras import backend as K

    return K.mean(z, axis=2)


input_a_sum = Lambda(f2)(input_a)
input_b_sum = Lambda(f2)(input_b)

output = Dense(1)(Dense(3, input_dim=4)(Concatenate()([input_a_sum, input_b_sum])))
model = Model(inputs=[input_a, input_b], outputs=output)
model.compile(loss="mean_squared_error", optimizer=SGD())
model.fit(
    [
        np.repeat(x.values[:, :2, np.newaxis], 3, axis=2),
        np.repeat(x.values[:, -2:, np.newaxis], 5, axis=2),
    ],
    y,
)

from tensorflow.keras.saving import get_custom_objects

global_custom_objects = get_custom_objects()

with open(os.path.join(save_path, "global_custom_objects.cloudpickle"), "wb") as out_f:
    cloudpickle.dump(global_custom_objects, out_f)

model_file_path = f"{save_path}/model.keras"
model.save(model_file_path)

```
then run shell command:

```
python tf-save.py .
```

It generates the following files in current directory:

```
global_custom_objects.cloudpickle  model.keras  model.weights.h5
```
One strange thing is it shouldn't generate `model.weights.h5` file. We only save model weights to `model.keras` file

then create a `tf-load.py` file containing:
```
import os.path
import sys
import cloudpickle
import tensorflow.keras

model_path = sys.argv[1]

custom_obj_path = os.path.join(model_path, "global_custom_objects.cloudpickle")
with open(custom_obj_path, "rb") as f:
    custom_objects = cloudpickle.load(f)

model_file_path = os.path.join(model_path, "model.keras")
tensorflow.keras.models.load_model(model_file_path, custom_objects=custom_objects)
```

then create a bash script `run.sh` like:

```
python tf-load.py . &
python tf-load.py . &
python tf-load.py . &
python tf-load.py . &

wait
```

then execute shell command
```
. run.sh 
```

error occurs:
```
Traceback (most recent call last):
  File "/tmp/tfm2/tf-load.py", line 13, in <module>
    tensorflow.keras.models.load_model(model_file_path, custom_objects=custom_objects)
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_api.py", line 182, in load_model
    return saving_lib.load_model(
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_lib.py", line 229, in load_model
    return _load_model_from_fileobj(
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/site-packages/keras/src/saving/saving_lib.py", line 353, in _load_model_from_fileobj
    weights_file_path.unlink()
  File "/home/weichen.xu/miniconda3/envs/mlflow/lib/python3.9/pathlib.py", line 1354, in unlink
    self._accessor.unlink(self)
FileNotFoundError: [Errno 2] No such file or directory: 'model.weights.h5'
```
and we found after executing `run.sh`, the `model.weights.h5` file is deleted.

**Comment by tilakrayal** [comment]
@WeichenXu123,
Weights can be saved to disk by calling **model.save_weights(filepath)**. The filename should end in **.weights.h5**. While loading the weights, Could you please try using **load_weights("my_model.weights.h5")**


```python
sequential_model.save_weights("my_model.weights.h5")
sequential_model.load_weights("my_model.weights.h5")
```

https://keras.io/guides/serialization_and_saving/#how-to-save-and-load-a-model

Thank you!

**Comment by WeichenXu123** [comment]
Thanks ! But I think `tensorflow.keras.models.load_model` should also work. In tensorflow released version it works.

And note that I saved the model as `.keras` format, I don't want to save it as `.weights.h5` format. 

The expected behavior should be:

`model.save("model_file.keras")` # generates a `model_file.keras` file.

`loaded_model = tensorflow.keras.models.load_model("model_file.keras")`

In released version above code works well. But now it becomes buggy.

**Comment by tongliang11** [comment]
I can reproduce the error. I tried python 3.10, and it won't generate "model.weights.h5" file, and loads well.

**Comment by tongliang11** [comment]
In my case, I got this error with my code (using mulitprocessing module in python to load keras model in each child process) running in a docker image on AWS. But I can't reproduce the error when running the same code locally. Very confusing.

**Comment by WeichenXu123** [comment]
The error occurs in Tensorflow 2.16.1 too. :)

**Comment by tilakrayal** [comment]
@WeichenXu123,
The default keras version in tensorflow v2.16.1 is the keras3.0 version which might be the reason for the change. While loading the weights,  please try using load_weights("my_model.weights.h5"). While tensorflow 2.15 version contains tf-keras 2.15(keras2). 

And for more Keras related issues, please feel free to raise the issue in  [keras-team/keras](https://github.com/keras-team/keras/issues) repo for quick resolution.
Thank you!

**Comment by WeichenXu123** [comment]
thanks!

**Comment by tilakrayal** [comment]
@WeichenXu123,
Could you please feel free to close this issue. Thank you!

**Comment by sachinprasadhs** [comment]
@WeichenXu123 , This issue has been fixed in the commit here https://github.com/keras-team/keras/pull/19924 and it will be available in the Keras 3.4.1 release which should be available tomorrow. 
Feel free to close the issue.
